pkg/util/machine/azureproviderspec.go

```diff
@@ -0,0 +1,52 @@
+package machine
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
+)
+
+type MachineType string
+
+const (
+	MachineSet MachineType = "machineset"
+	Machine    MachineType = "machine"
+)
+
+// UnmarshalAzureProviderSpec unmarshals an Azure Provider Spec set on machines or machinesets
+// it contains backward compatibility for the older azureproviderconfig which is no longer in use
+// in OCP 4.10+
+func UnmarshalAzureProviderSpec(name string, mType MachineType, rawProviderSpec []byte) (*machinev1beta1.AzureMachineProviderSpec, error) {
+	var machineProviderSpec *machinev1beta1.AzureMachineProviderSpec
+
+	if strings.Contains(string(rawProviderSpec), "azureproviderconfig.openshift.io") {
+		machineProviderSpec = &machinev1beta1.AzureMachineProviderSpec{}
+		err := json.Unmarshal(rawProviderSpec, machineProviderSpec)
+		if err != nil {
+			return machineProviderSpec, fmt.Errorf("%s %s: failed to unmarshal the 'azureproviderconfig.openshift.io' provider spec: %q", mType, name, err.Error())
+		}
+	} else {
+		o, _, err := scheme.Codecs.UniversalDeserializer().Decode(rawProviderSpec, nil, nil)
+		if err != nil {
+			return machineProviderSpec, err
+		}
+
+		var ok bool
+		machineProviderSpec, ok = o.(*machinev1beta1.AzureMachineProviderSpec)
+		if !ok {
+			// If this happens, the azure machine provider spec type/apiversion may have been updated and
+			// we need to handle it appropriately
+			return machineProviderSpec, fmt.Errorf("%s %s: failed to read provider spec: %T", mType, name, o)
+		}
+	}
+
+	return machineProviderSpec, nil
```

Comment:
Couldn't we get rid of the `var machineProviderSpec` and instantiate it inside the `if` ? That would also get rid of the `var ok bool` since we could just use `:=` on line 43. It would require to put the return inside the branches, but I think that's a little better. 

Acutally, what could make it even clearer would be to have functions for each use case: something like `unmarshalLegacy` and `unmarshalAMPS` which we could just return. That would make it clearer which is the legacy and which is the current

Reply:
This is a flow starts with a variable that we want to return and ultimately returns that variable with the logic required in between. I can see the use case for 2 distinct functions, but the logic is not verbose and easy to follow so I think there are diminishing returns by refactoring them into individual functions.

Reply:
it is a little confusing though, if you look at line 39, we return `machineProviderSpec` because it has been defined before the if statement, but what we really want to return is `nil`, which is what the variable is 100% of the time. It would not happen if we returned straight from the branches. 

If we had something like this, it would be clearer. 
```go 
if something {
    return resultForSomething
} else {
    return defaultResult
}
``` 

But we should at least have a little comment in the if branches to say which one is the legacy and which one is the current

Reply:
I do agree that we should be returning an explicit `nil` in those error blocks to be clearer of the return value.

Reply:
btw, I we could even get rid of the else. Unnecessary elses are usually avoided (cf effective go + uber go style guide)
```suggestion
	var machineProviderSpec *machinev1beta1.AzureMachineProviderSpec

	if strings.Contains(string(rawProviderSpec), "azureproviderconfig.openshift.io") {
		machineProviderSpec = &machinev1beta1.AzureMachineProviderSpec{}
		err := json.Unmarshal(rawProviderSpec, machineProviderSpec)
		if err != nil {
			return nil, fmt.Errorf("%s %s: failed to unmarshal the 'azureproviderconfig.openshift.io' provider spec: %q", mType, name, err.Error())
		}
		return  machineProviderSpec, nil
	} 
	
	o, _, err := scheme.Codecs.UniversalDeserializer().Decode(rawProviderSpec, nil, nil)
	if err != nil {
		return nil, err
	}

	var ok bool
	machineProviderSpec, ok = o.(*machinev1beta1.AzureMachineProviderSpec)
	if !ok {
		// If this happens, the azure machine provider spec type/apiversion may have been updated and
		// we need to handle it appropriately
		return nil, fmt.Errorf("%s %s: failed to read provider spec: %T", mType, name, o)
	}
	

	return machineProviderSpec, nil
```

Reply:
Thanks for pointing out the simplification, I'll remove the extraneous else and update to explicitly return nil when appropriate.